_episodes/07-thresholding.md

```diff
@@ -14,532 +14,370 @@ threshold an image represented by a numpy array."
-thresholding, just as we did in the previous section. Since the program is 
-intended to produce numeric output, without a person shepherding it, it does
-not display any of the images.
+    # determine root mass ratio
+    rootPixels = np.count_nonzero(binary_mask)
+    w = binary_mask.shape[1]
+    h = binary_mask.shape[0]
+    density = rootPixels / (w * h)
 
-~~~
-# perform adaptive thresholding to produce a binary image
-t = skimage.filters.threshold_otsu(blur)
-binary = blur > t
+    return density
 ~~~
 {: .language-python}
 
-We do, however, want to save the binary images, in case we wish to examine them
-at a later time. That is what this block of code does:
-~~~
-# save binary image; first find beginning of file extension
-dot = filename.index(".")
-binary_file_name = filename[:dot] + "-binary" + filename[dot:]
-skimage.io.imsave(fname=binary_file_name, arr=skimage.img_as_ubyte(binary))
-~~~
-{: .language-python}
+The function begins with reading the orignal image from the file `filename`. We use `skimage.io.imread` with the optional argument `as_gray=True` to automatically convert it to grayscale. Next, the grayscale image is blurred with a Gaussian filter with the value of `sigma` that is passed to the function. Then we determine the threshold `t` with Otsu's method and create a binary mask just as we did in the previous section. Up to this point, everything should be familiar.
 
-This code does a little bit of string manipulation to determine the filename 
-to use when the binary image is saved. For example, if the input filename being
-processed is **trial-020.jpg**, we want to save the corresponding binary image
-as **trial-020-binary.jpg**. To do that, we first determine the index of the 
-dot between the filename and extension -- and note that we assume that there is
-only one dot in the filename! Once we have the location of the dot, we can use
-slicing to pull apart the filename string, inserting "-binary" in between the
-end of the original name and the extension. Then, the binary image is saved via
-a call to the `skimage.io.imsave()` function. 
-In order to convert from the binary range of 0 and 1 of the mask to a gray level image that can be saved as png, we use the `skimage.img_as_ubyte` utility function.
+The final part of the function determines the root mass ratio in the image. Recall that in the `binary_mask`, every pixel has either a value of zero (black/background) or one (white/foreground). We want to count the number of white pixels, which can be accomplished with a call to the numpy function `np.count_nonzero`. Then we determine the width and height of the image by using the the elements of `binary_mask.shape` (that is, the dimensions of the numpy array that stores the image). Finally, the density ratio is calculated by dividing the number of white pixles by the total number of pixels `w*h` in the image. The function returns then root density of the image.
```

Comment:
> We want to count the number of white pixels, which can be accomplished with a call to the numpy function `np.count_nonzero`.

I am curious about the reasoning behind using `np.count_nonzero` instead of a summing function here? If the image is binary, the result should be the same. Is `count_nonzero` a more universally-applicable function? Or somehow more convenient than `sum` or `np.sum` in this case?

Reply:
Thanks for bringing this up @tobyhodges. To be honest, I adopted the code from the previous version which used `np.count_nonzero` and mentioned it in the learning objectives. I think an argument could be made that the operation involves counting pixels rather than summing pixel values. The counting could be generalized to pixels above or below a threshold, although this would involve other function calls. I have no preference on the use of `np.count_nonzero` vs `np.sum` and would welcome any additional comments.

This prompts a more general question on how much "new" NumPy functionality we want to use in the lesson, which I think would be good to discuss in an upcoming co-working session (see also my [comment](https://github.com/datacarpentry/image-processing/pull/156#issuecomment-907690103) in #156).

Reply:
I have opened #159 to discuss in co-working session.